File: articles/participatory-modeling-and-stakeholder-systems/r/participatory_modeling_workflow.R
# participatory_modeling_workflow.R
# Base R workflow:
# stakeholder scenario scoring and consensus diagnostics.

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)

if (length(file_arg) > 0) {
  script_path <- normalizePath(sub("^--file=", "", file_arg[1]), mustWork = TRUE)
  article_root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)
} else {
  article_root <- normalizePath(getwd(), mustWork = TRUE)
}

data_dir <- file.path(article_root, "data")
tables_dir <- file.path(article_root, "outputs", "tables")
figures_dir <- file.path(article_root, "outputs", "figures")

dir.create(tables_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(figures_dir, recursive = TRUE, showWarnings = FALSE)

stakeholders <- read.csv(file.path(data_dir, "stakeholder_weights.csv"), stringsAsFactors = FALSE)
scenarios <- read.csv(file.path(data_dir, "scenarios.csv"), stringsAsFactors = FALSE)
assumptions <- read.csv(file.path(data_dir, "assumption_register.csv"), stringsAsFactors = FALSE)

outcomes <- c("access", "cost", "resilience", "equity", "feasibility")

score_rows <- data.frame()

for (i in seq_len(nrow(stakeholders))) {
  for (j in seq_len(nrow(scenarios))) {
    score <- sum(as.numeric(stakeholders[i, outcomes]) * as.numeric(scenarios[j, outcomes]))

    score_rows <- rbind(
      score_rows,
      data.frame(
        stakeholder_group = stakeholders$stakeholder_group[i],
        scenario = scenarios$scenario[j],
        score = score
      )
    )
  }
}

scenario_summary <- aggregate(
  score ~ scenario,
  data = score_rows,
  FUN = function(x) c(mean = mean(x), sd = sd(x), min = min(x), max = max(x))
)

scenario_summary <- do.call(data.frame, scenario_summary)
names(scenario_summary) <- c(
  "scenario",
  "mean_score",
  "disagreement_sd",
  "minimum_score",
  "maximum_score"
)

scenario_summary$score_range <- scenario_summary$maximum_score - scenario_summary$minimum_score

lambda <- 0.50
scenario_summary$legitimacy_adjusted_score <-
  scenario_summary$mean_score - lambda * scenario_summary$disagreement_sd

scenario_summary$consensus_label <- ifelse(
  scenario_summary$disagreement_sd >= 0.08,
  "high disagreement",
  ifelse(scenario_summary$disagreement_sd >= 0.04, "moderate disagreement", "low disagreement")
)

scenario_summary <- scenario_summary[order(-scenario_summary$legitimacy_adjusted_score), ]

assumption_summary <- aggregate(
  assumption_id ~ status,
  data = assumptions,
  FUN = length
)

names(assumption_summary) <- c("status", "assumption_count")

validation_checks <- data.frame(
  check = c(
    "stakeholder_weights_sum_to_one",
    "scenario_scores_between_zero_and_one",
    "scenario_summary_created",
    "assumption_register_created"
  ),
  passed = c(
    all(abs(rowSums(stakeholders[, outcomes]) - 1) < 1e-9),
    all(score_rows$score >= 0 & score_rows$score <= 1),
    nrow(scenario_summary) > 0,
    nrow(assumptions) > 0
  )
)

write.csv(
  stakeholders,
  file.path(tables_dir, "r_participatory_stakeholder_weights.csv"),
  row.names = FALSE
)

write.csv(
  scenarios,
  file.path(tables_dir, "r_participatory_scenarios.csv"),
  row.names = FALSE
)

write.csv(
  score_rows,
  file.path(tables_dir, "r_participatory_stakeholder_scenario_scores.csv"),
  row.names = FALSE
)

write.csv(
  scenario_summary,
  file.path(tables_dir, "r_participatory_scenario_summary.csv"),
  row.names = FALSE
)

write.csv(
  assumptions,
  file.path(tables_dir, "r_participatory_assumption_register.csv"),
  row.names = FALSE
)

write.csv(
  assumption_summary,
  file.path(tables_dir, "r_participatory_assumption_status_summary.csv"),
  row.names = FALSE
)

write.csv(
  validation_checks,
  file.path(tables_dir, "r_participatory_validation_checks.csv"),
  row.names = FALSE
)

png(file.path(figures_dir, "r_participatory_scenario_scores.png"), width = 1000, height = 700)
barplot(
  scenario_summary$legitimacy_adjusted_score,
  names.arg = scenario_summary$scenario,
  las = 2,
  ylab = "Legitimacy-Adjusted Score",
  main = "Participatory Scenario Comparison"
)
grid()
dev.off()

print(scenario_summary)
print(validation_checks)
cat("R participatory modeling workflow complete.\n")

File: articles/participatory-modeling-and-stakeholder-systems/python/participatory_modeling_workflow.py
#!/usr/bin/env python3
"""
Participatory modeling and stakeholder systems workflow.

Dependency-light workflow demonstrating:

1. Stakeholder group definitions
2. Outcome weights
3. Scenario performance values
4. Stakeholder-specific scenario scoring
5. Disagreement diagnostics
6. Legitimacy-adjusted rankings
7. Assumption and governance registers
8. Validation checks

All data are synthetic.
"""

from __future__ import annotations

from pathlib import Path
import csv
import math
from statistics import mean, pstdev


ARTICLE_ROOT = Path(__file__).resolve().parents[1]
DATA = ARTICLE_ROOT / "data"
TABLES = ARTICLE_ROOT / "outputs" / "tables"


OUTCOMES = ["access", "cost", "resilience", "equity", "feasibility"]


def read_csv(path: Path) -> list[dict[str, str]]:
    with path.open("r", newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def write_csv(path: Path, rows: list[dict[str, object]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    if not rows:
        raise ValueError(f"No rows to write: {path}")

    fieldnames: list[str] = []
    for row in rows:
        for key in row:
            if key not in fieldnames:
                fieldnames.append(key)

    with path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=fieldnames, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(rows)


def stakeholder_weight_sum(stakeholder: dict[str, str]) -> float:
    return sum(float(stakeholder[outcome]) for outcome in OUTCOMES)


def score_scenarios(
    stakeholders: list[dict[str, str]],
    scenarios: list[dict[str, str]],
) -> list[dict[str, object]]:
    rows: list[dict[str, object]] = []

    for stakeholder in stakeholders:
        for scenario in scenarios:
            score = sum(float(stakeholder[outcome]) * float(scenario[outcome]) for outcome in OUTCOMES)

            rows.append({
                "stakeholder_group": stakeholder["stakeholder_group"],
                "scenario": scenario["scenario"],
                "score": round(score, 6),
            })

    return rows


def summarize_scores(score_rows: list[dict[str, object]], scenarios: list[dict[str, str]]) -> list[dict[str, object]]:
    summary_rows: list[dict[str, object]] = []

    for scenario in scenarios:
        scenario_name = scenario["scenario"]
        scores = [
            float(row["score"])
            for row in score_rows
            if row["scenario"] == scenario_name
        ]

        mean_score = mean(scores)
        disagreement_sd = pstdev(scores)
        minimum_score = min(scores)
        maximum_score = max(scores)
        legitimacy_adjusted_score = mean_score - 0.50 * disagreement_sd

        if disagreement_sd >= 0.08:
            consensus_label = "high disagreement"
        elif disagreement_sd >= 0.04:
            consensus_label = "moderate disagreement"
        else:
            consensus_label = "low disagreement"

        summary_rows.append({
            "scenario": scenario_name,
            "mean_score": round(mean_score, 6),
            "disagreement_sd": round(disagreement_sd, 6),
            "minimum_score": round(minimum_score, 6),
            "maximum_score": round(maximum_score, 6),
            "score_range": round(maximum_score - minimum_score, 6),
            "legitimacy_adjusted_score": round(legitimacy_adjusted_score, 6),
            "consensus_label": consensus_label,
        })

    summary_rows.sort(key=lambda row: float(row["legitimacy_adjusted_score"]), reverse=True)
    return summary_rows


def assumption_status_summary(assumptions: list[dict[str, str]]) -> list[dict[str, object]]:
    counts: dict[str, int] = {}
    for row in assumptions:
        counts[row["status"]] = counts.get(row["status"], 0) + 1

    return [
        {
            "status": status,
            "assumption_count": count,
        }
        for status, count in sorted(counts.items())
    ]


def main() -> None:
    stakeholders = read_csv(DATA / "stakeholder_weights.csv")
    scenarios = read_csv(DATA / "scenarios.csv")
    assumptions = read_csv(DATA / "assumption_register.csv")

    score_rows = score_scenarios(stakeholders, scenarios)
    summary_rows = summarize_scores(score_rows, scenarios)
    assumption_summary = assumption_status_summary(assumptions)

    validation_rows = [
        {
            "check": "stakeholder_weights_sum_to_one",
            "passed": all(math.isclose(stakeholder_weight_sum(stakeholder), 1.0, abs_tol=1e-9) for stakeholder in stakeholders),
            "value": "all_groups_checked",
        },
        {
            "check": "scenario_scores_between_zero_and_one",
            "passed": all(0.0 <= float(row["score"]) <= 1.0 for row in score_rows),
            "value": "all_scores_checked",
        },
        {
            "check": "scenario_summary_created",
            "passed": len(summary_rows) == len(scenarios),
            "value": len(summary_rows),
        },
        {
            "check": "assumption_register_created",
            "passed": len(assumptions) > 0,
            "value": len(assumptions),
        },
        {
            "check": "participation_levels_exported",
            "passed": len(read_csv(DATA / "participation_levels.csv")) > 0,
            "value": "participation_levels_available",
        },
    ]

    write_csv(TABLES / "python_participatory_stakeholder_weights.csv", stakeholders)
    write_csv(TABLES / "python_participatory_scenarios.csv", scenarios)
    write_csv(TABLES / "python_participatory_stakeholder_scenario_scores.csv", score_rows)
    write_csv(TABLES / "python_participatory_scenario_summary.csv", summary_rows)
    write_csv(TABLES / "python_participatory_assumption_register.csv", assumptions)
    write_csv(TABLES / "python_participatory_assumption_status_summary.csv", assumption_summary)
    write_csv(TABLES / "python_participation_levels.csv", read_csv(DATA / "participation_levels.csv"))
    write_csv(TABLES / "python_facilitation_risks.csv", read_csv(DATA / "facilitation_risks.csv"))
    write_csv(TABLES / "python_evidence_sources.csv", read_csv(DATA / "evidence_sources.csv"))
    write_csv(TABLES / "python_modeling_approaches.csv", read_csv(DATA / "modeling_approaches.csv"))
    write_csv(TABLES / "python_participatory_validation_checks.csv", validation_rows)

    print("Participatory modeling workflow complete.")
    print(TABLES / "python_participatory_scenario_summary.csv")


if __name__ == "__main__":
    main()
